// react-native-navigation/lib/dist/interfaces/Options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OptionsModalPresentationStyle;
(function (OptionsModalPresentationStyle) {
    OptionsModalPresentationStyle["formSheet"] = "formSheet";
    OptionsModalPresentationStyle["pageSheet"] = "pageSheet";
    OptionsModalPresentationStyle["overFullScreen"] = "overFullScreen";
    OptionsModalPresentationStyle["overCurrentContext"] = "overCurrentContext";
    OptionsModalPresentationStyle["currentContext"] = "currentContext";
    OptionsModalPresentationStyle["popOver"] = "popOver";
    OptionsModalPresentationStyle["fullScreen"] = "fullScreen";
    OptionsModalPresentationStyle["none"] = "none";
})(OptionsModalPresentationStyle = exports.OptionsModalPresentationStyle || (exports.OptionsModalPresentationStyle = {}));
var OptionsModalTransitionStyle;
(function (OptionsModalTransitionStyle) {
    OptionsModalTransitionStyle["coverVertical"] = "coverVertical";
    OptionsModalTransitionStyle["crossDissolve"] = "crossDissolve";
    OptionsModalTransitionStyle["flipHorizontal"] = "flipHorizontal";
    OptionsModalTransitionStyle["partialCurl"] = "partialCurl";
})(OptionsModalTransitionStyle = exports.OptionsModalTransitionStyle || (exports.OptionsModalTransitionStyle = {}));
